### PR TITLE
[AIRFLOW-6327] http_hook: Accept json= parameter for payload

### PR DESCRIPTION
**Pull request by baolsen**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6327
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Add json parameter to http_hook to allow easier usage when posting Python objects. See JIRA for more info: https://issues.apache.org/jira/browse/AIRFLOW-6327

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Relevant test added

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/hooks/http_hook.py [hunk 1]
```diff
@@ -75,20 +75,24 @@ def get_conn(self, headers=None):
                 try:
                     session.headers.update(conn.extra_dejson)
                 except TypeError:
-                    self.log.warning('Connection to %s has invalid extra field.', conn.host)
+                    self.log.warning(
+                        'Connection to %s has invalid extra field.', conn.host)
```
Comment (zhongjiajie) [inline airflow/hooks/http_hook.py]: I think this line is unrelated and unnecessary change? So as somewhere else in this PR

File: airflow/hooks/http_hook.py [hunk 2]
```diff
@@ -75,20 +75,24 @@ def get_conn(self, headers=None):
...
+    def run(self, endpoint, data=None, json=None, headers=None, extra_options=None):
         """
         Performs the request
 
         :param endpoint: the endpoint to be called i.e. resource/v1/query?
         :type endpoint: str
         :param data: payload to be uploaded or request parameters
         :type data: dict
+        :param json: Python object which is encoded and posted as JSON. Usually
+          this is a dict but can be any JSON-encodable Python object.
```
Comment (zhongjiajie) [inline airflow/hooks/http_hook.py]: Four blank space will be better.

**Review by zhongjiajie — COMMENTED** [review]
My 2 cents

File: airflow/hooks/http_hook.py [hunk 3]
```diff
@@ -175,7 +181,8 @@ def run_and_check(self, session, prepped_request, extra_options):
             return response
 
         except requests.exceptions.ConnectionError as ex:
-            self.log.warning(str(ex) + ' Tenacity will retry to execute the operation')
+            self.log.warning(
```
Comment (dstandish) [inline airflow/hooks/http_hook.py]: as @zhongjiajie has stated, you are introducing line breaks unnecessarily, which makes the diff unnecessarily noisy (and the lines were not overly long to begin with)
currently flake8 is set at 110 characters, so probably best to set it to 110 in your editor, and these extraneous line breaks.
see https://github.com/apache/airflow/blob/master/.flake8#L2

File: airflow/hooks/http_hook.py [hunk 4]
```diff
@@ -123,6 +127,7 @@ def run(self, endpoint, data=None, headers=None, extra_options=None):
             req = requests.Request(self.method,
                                    url,
                                    data=data,
+                                   json=json,
```
Comment (dstandish) [inline airflow/hooks/http_hook.py]: i see you are only adding `json` for requests other than GET or HEAD

i know that presently the hook conditionally forwards the `data` param --  in one case to `params`, in another to `data`.  but this seems like a mistake.  

since `json` is brand new param, why not pass it on in all cases, no matter the request type?

at least, the docstring should probably disclose that `json` is only forwarded for non-GET and non-HEAD requests, if there is a legitimate reason to forward it selectively.  and perhaps raise an error if it is supplied and you will not forward it.

File: tests/hooks/test_http_hook.py [hunk 5]
```diff
@@ -334,5 +338,35 @@ def test_connection_without_host(self, mock_get_connection):
...
+            additional_matcher=match_obj1
+        )
+
+        with mock.patch(
+            'airflow.hooks.base_hook.BaseHook.get_connection',
+            side_effect=get_airflow_connection
+        ):
+            # Send obj1 as JSON and verify it matched the mock
+            resp = self.post_hook.run('v1/test', json=obj1)
+            self.assertEqual(resp.status_code, 200)
```
Comment (dstandish) [inline tests/hooks/test_http_hook.py]: test guidelines suggest we should use standard asserts now 
https://github.com/apache/airflow/blob/master/TESTING.rst#writing-unit-tests

File: tests/hooks/test_http_hook.py [hunk 6]
```diff
@@ -334,5 +338,35 @@ def test_connection_without_host(self, mock_get_connection):
...
+            self.assertEqual(resp.text, 'test_post_json_request')
+        # Ensure that obj1 was what was sent
```
Comment (dstandish) [inline tests/hooks/test_http_hook.py]: this comment does not seem accurate

File: tests/hooks/test_http_hook.py [hunk 7]
```diff
@@ -334,5 +338,35 @@ def test_connection_without_host(self, mock_get_connection):
...
+        with self.assertRaises(requests_mock.exceptions.NoMockAddress):
+            resp = self.post_hook.run('v1/test', json=obj2)
```
Comment (dstandish) [inline tests/hooks/test_http_hook.py]: seems you are just verifying here that the request mock raises with bad address...  does this belong in `test_post_json_request`?

**Review by dstandish — CHANGES_REQUESTED** [review]

File: airflow/hooks/http_hook.py (hunk 1 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: Thanks, I'll remove them

File: airflow/hooks/http_hook.py (hunk 2 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: Thanks, will fix

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: Thanks, Python requests library only has data= and json= parameters for the requests.post method. I'll update the docstring to be clearer

File: airflow/hooks/http_hook.py (hunk 3 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: Thanks, I've fixed it in my editor

File: tests/hooks/test_http_hook.py (hunk 5 above)
Comment (baolsen) [reply tests/hooks/test_http_hook.py]: Thanks, will update

File: tests/hooks/test_http_hook.py (hunk 6 above)
Comment (baolsen) [reply tests/hooks/test_http_hook.py]: Comment clarified, thanks

File: tests/hooks/test_http_hook.py (hunk 7 above)
Comment (baolsen) [reply tests/hooks/test_http_hook.py]: I wanted to be sure that obj1 was encoded & posted correctly, not just that a post was done. 
So I needed to check the contents of the request payload - using additional_matcher to ensure that what was posted matched. However since request_mock is a bit confusing to me I wasnt convinced that additional_matcher was going to be used. So I added a check to post obj2 the same way as obj1, which raises an exception

**Comment by baolsen** [comment]
Hi @zhongjiajie and @dstandish , ready for your review again :)

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (dstandish) [reply airflow/hooks/http_hook.py]: Though `json` is not a named argument in `get`, it does work.  Kwargs are passed on to `request.  See [here](https://github.com/psf/requests/blob/master/requests/api.py#L76).

My inclination would be to just blindly forward all args (param / data / json) in all cases, on the basis that we should not get in the way between user and library, without some reason.  But you don't have to agree with me :)

File: tests/hooks/test_http_hook.py (hunk 7 above)
Comment (dstandish) [reply tests/hooks/test_http_hook.py]: ok i was confused by the name `NoMockAddress`.  it's a little confusing, but i now understand that's how requests mock is designed.

but anyway, i think you should chop this.  i'd just test your feature, and not the functionality of requests-mock.

in any case, i don't think this bit is doing what you want it to do.  on line 368 if you replace `obj2` with `obj1`, the test still passes.  this is because you aren't mocking `airflow.hooks.base_hook.BaseHook.get_connection` at this point of the code, so it is finding no match, even when you supply the correct object.

File: tests/hooks/test_http_hook.py [hunk 8]
```diff
@@ -334,5 +335,37 @@ def test_connection_without_host(self, mock_get_connection):
...
         self.assertEqual(hook.base_url, 'http://')
 
+    @requests_mock.mock()
+    def test_post_json_request(self, mock_requests):
+        obj1 = {'a': 1, 'b': 'abc', 'c': [1, 2, {"d": 10}]}
+        obj2 = [1, 2, 3]
+
+        # Ensure that obj1 was encoded to JSON
+        def match_obj1(request):
+            return json.loads(request.text) == obj1
```
Comment (dstandish) [inline tests/hooks/test_http_hook.py]:
```suggestion
            return request.json() == obj1
```

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: Ah I didn't see that. Thanks for pointing it out.
I'm also in favor of being able to forward any kwargs down to the library, will see what I can come up with.

File: tests/hooks/test_http_hook.py (hunk 7 above)
Comment (baolsen) [reply tests/hooks/test_http_hook.py]: Interesting, thanks for testing that. Requests_mock is tricky to get my head around. Anyway the second test is a bit overkill and if it isn't doing what I thought then it's certainly worth removing it :)

File: tests/hooks/test_http_hook.py (hunk 8 above)
Comment (baolsen) [reply tests/hooks/test_http_hook.py]: Thanks, I've changed per your suggestion.

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: @dstandish I've changed my approach a bit - 
Now you'll see "**request_kwargs" which is more future proof. Another benefit is I could remove the changes to the slack_webhook tests. 
Would love to hear your thoughts :) I tried to add a test for the other kwargs supported by Requests but didn't manage to get it to work properly, and decided to leave it just testing the json= kwarg.
Edit: Also finally figured out how rebase works, and cleaned up the commit history.

**Review by turbaszek — APPROVED** [review]

File: tests/hooks/test_http_hook.py [hunk 9]
```diff
@@ -334,5 +334,42 @@ def test_connection_without_host(self, mock_get_connection):
         hook.get_conn({})
         self.assertEqual(hook.base_url, 'http://')
 
+    @requests_mock.mock()
```
Comment (dstandish) [inline tests/hooks/test_http_hook.py:342]: this is may be super nit picky, but you could simplify the test using parameters if you want to

```suggestion
    @parameterized.expand([
        'GET',
        'POST',
    ])
    @requests_mock.mock()
    def test_json_request(self, method, mock_requests):
        obj1 = {'a': 1, 'b': 'abc', 'c': [1, 2, {"d": 10}]}

        def match_obj1(request):
            return request.json() == obj1

        mock_requests.request(
            method=method,
            url='//test:8080/v1/test',
            additional_matcher=match_obj1
        )

        with mock.patch(
            'airflow.hooks.base_hook.BaseHook.get_connection',
            side_effect=get_airflow_connection
        ):
            # will raise NoMockAddress exception if obj1 != request.json()
            HttpHook(method=method).run('v1/test', json=obj1)
```

here i am also suggesting removing those asserts.

they are just checking that your mock is returning what you told it to return -- the fact that no exception is raised in `hook.run`  is the actual indicator that the test was passed.  the asserts don't test anything beyond that.

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (dstandish) [reply airflow/hooks/http_hook.py]: hi @baolsen i like your `request_kwargs` approach.

i just offered a suggestion for cleaning up the test.  you'll also have to add `from parameterized import parameterized` if you choose to adopt the simplification

**Review by zhongjiajie — APPROVED** [review]

File: airflow/hooks/http_hook.py (hunk 4 above)
Comment (baolsen) [reply airflow/hooks/http_hook.py]: That's neat, I hadn't seen that before. Thanks :) I've included your suggested changes.
